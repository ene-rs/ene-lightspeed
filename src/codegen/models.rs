use crate::models::entity::{Entity, FilterByAttributes, MostSpecificFilterAttributes};

use super::CodeGen;
use heck::{
    ToKebabCase, ToLowerCamelCase, ToShoutyKebabCase, ToShoutySnakeCase, ToSnakeCase, ToTitleCase,
    ToTrainCase, ToUpperCamelCase,
};
#[derive(Debug, Clone)]
pub struct EntityTemplate {
    pub entity_content: Vec<EntityTemplateContent>,
}

impl CodeGen for EntityTemplate {
    fn generate_code(&self, entity: &Entity, entities: &Vec<Entity>) -> anyhow::Result<String> {
        self.entity_content
            .iter()
            .map(|entity_content| entity_content.generate_code(entity, entities))
            .collect::<anyhow::Result<Vec<String>>>()
            .map(|entity_contents| entity_contents.join(""))
    }
}

#[derive(Debug, Clone)]
pub enum Content {
    Static(String),
    EntityName(StringWithNamingConvention),
    EntityPlural(StringWithNamingConvention),
    PrimaryKey(StringWithNamingConvention),
}

impl CodeGen for Content {
    fn generate_code(&self, entity: &Entity, _entities: &Vec<Entity>) -> anyhow::Result<String> {
        Ok(match self {
            Self::Static(value) => value.clone(),
            Self::EntityName(naming_convention) => {
                naming_convention.to_naming_convention(&entity.name)
            }
            Self::EntityPlural(naming_convention) => {
                naming_convention.to_naming_convention(&entity.semantics.plural)
            }
            Self::PrimaryKey(naming_convention) => {
                naming_convention.to_naming_convention(&entity.primary_key)
            }
        })
    }
}

#[derive(Debug, Clone)]
pub enum EntityTemplateContent {
    Content(Content),
    AttributeRep(AttributeRep),
    FilterRep(FilterRep),
}

impl CodeGen for EntityTemplateContent {
    fn generate_code(&self, entity: &Entity, entities: &Vec<Entity>) -> anyhow::Result<String> {
        match self {
            Self::Content(content) => content.generate_code(entity, entities),
            Self::AttributeRep(attribute_rep) => attribute_rep.generate_code(entity, entities),
            Self::FilterRep(filter_rep) => filter_rep.generate_code(entity, entities),
        }
    }
}

#[derive(Debug, Clone)]
pub enum AttributeRep {
    NamesAndTypesPresent {
        prefix: Content,
        template_attribute_name: StringWithNamingConvention,
        attribute_name_and_type_separator: Content,
        template_attribute_type: StringWithNamingConvention,
        attributes_separator: Content,
        suffix: Content,
    },
    NamesPresent {
        prefix: Content,
        template_attribute_name: StringWithNamingConvention,
        attributes_separator: Content,
        suffix: Content,
    },
    NamesTwicePresent {
        prefix: Content,
        template_attribute_name: StringWithNamingConvention,
        attribute_name_twice_separator: Content,
        attributes_separator: Content,
        suffix: Content,
    },
}

impl CodeGen for AttributeRep {
    fn generate_code(&self, entity: &Entity, entities: &Vec<Entity>) -> anyhow::Result<String> {
        match self {
            Self::NamesAndTypesPresent {
                prefix,
                template_attribute_name,
                attribute_name_and_type_separator,
                template_attribute_type,
                attributes_separator,
                suffix,
            } => {
                let prefix = prefix.generate_code(entity, entities)?;
                let attributes = entity
                    .attributes
                    .iter()
                    .map(
                        |(entity_attribute_name, entity_attribute_type)| -> anyhow::Result<String> {
                            let attribute_name_and_type_separator =
                                attribute_name_and_type_separator
                                    .generate_code(entity, entities)?;
                            let attribute_type = template_attribute_type.to_naming_convention(
                                &entity_attribute_type.get_attribute_type(entities)?,
                            );
                            Ok(format!(
                                "{}{}{}",
                                template_attribute_name.to_naming_convention(entity_attribute_name),
                                attribute_name_and_type_separator,
                                attribute_type
                            ))
                        },
                    )
                    .collect::<anyhow::Result<Vec<String>>>()?
                    .join(
                        attributes_separator
                            .generate_code(entity, entities)?
                            .as_str(),
                    );
                let suffix = suffix.generate_code(entity, entities)?;
                Ok(format!("{}{}{}", prefix, attributes, suffix))
            }
            Self::NamesPresent {
                prefix,
                template_attribute_name,
                attributes_separator,
                suffix,
            } => {
                let prefix = prefix.generate_code(entity, entities)?;
                let attributes = entity
                    .attributes
                    .iter()
                    .map(|(entity_attribute_name, _)| -> anyhow::Result<String> {
                        Ok(template_attribute_name.to_naming_convention(entity_attribute_name))
                    })
                    .collect::<anyhow::Result<Vec<String>>>()?
                    .join(
                        attributes_separator
                            .generate_code(entity, entities)?
                            .as_str(),
                    );
                let suffix = suffix.generate_code(entity, entities)?;
                Ok(format!("{}{}{}", prefix, attributes, suffix))
            }
            Self::NamesTwicePresent {
                prefix,
                template_attribute_name,
                attribute_name_twice_separator,
                attributes_separator,
                suffix,
            } => {
                let prefix = prefix.generate_code(entity, entities)?;
                let attributes = entity
                    .attributes
                    .iter()
                    .map(|(entity_attribute_name, _)| -> anyhow::Result<String> {
                        Ok(format!(
                            "{}{}{}",
                            template_attribute_name.to_naming_convention(entity_attribute_name),
                            attribute_name_twice_separator.generate_code(entity, entities)?,
                            template_attribute_name.to_naming_convention(entity_attribute_name)
                        ))
                    })
                    .collect::<anyhow::Result<Vec<String>>>()?
                    .join(
                        attributes_separator
                            .generate_code(entity, entities)?
                            .as_str(),
                    );
                let suffix = suffix.generate_code(entity, entities)?;
                Ok(format!("{}{}{}", prefix, attributes, suffix))
            }
        }
    }
}

#[derive(Debug, Clone)]
pub enum ContentWithMSFA {
    Content(Content),
    MostSpecificFilterAttributeName(StringWithNamingConvention),
    MostSpecificFilterAttributeType(StringWithNamingConvention),
}

impl ContentWithMSFA {
    pub fn generate_code(
        &self,
        filter_by_attributes: &FilterByAttributes,
        entity: &Entity,
        entities: &Vec<Entity>,
    ) -> anyhow::Result<String> {
        match self {
            Self::Content(content) => content.generate_code(entity, entities),
            Self::MostSpecificFilterAttributeName(naming_convention) => {
                let (most_specific_filter_attribute_name, _) =
                    filter_by_attributes.get_most_specific_filter_attribute();
                Ok(naming_convention.to_naming_convention(&most_specific_filter_attribute_name))
            }
            Self::MostSpecificFilterAttributeType(naming_convention) => {
                let (_, most_specific_filter_attribute_type) =
                    filter_by_attributes.get_most_specific_filter_attribute();
                Ok(naming_convention.to_naming_convention(
                    &most_specific_filter_attribute_type.get_attribute_type(entities)?,
                ))
            }
        }
    }
}

#[derive(Debug, Clone)]
pub enum FilterRep {
    ConsecutiveFilters { filter: Filter },
    SeparatedFilters { filter: Filter, separator: Content },
}

impl CodeGen for FilterRep {
    fn generate_code(&self, entity: &Entity, entities: &Vec<Entity>) -> anyhow::Result<String> {
        match self {
            Self::ConsecutiveFilters { filter } => entity
                .get_filter_by_attributes()
                .iter()
                .map(|filter_by_attributes| {
                    filter.generate_code(filter_by_attributes, entity, entities)
                })
                .collect::<anyhow::Result<Vec<String>>>()
                .map(|filters| filters.join("")),
            Self::SeparatedFilters { filter, separator } => entity
                .get_filter_by_attributes()
                .iter()
                .map(|filter_by_attributes| {
                    filter.generate_code(filter_by_attributes, entity, entities)
                })
                .collect::<anyhow::Result<Vec<String>>>()
                .map(|filters| -> anyhow::Result<String> {
                    Ok(filters.join(separator.generate_code(entity, entities)?.as_str()))
                })?,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Filter {
    pub content: Vec<FilterBody>,
}

impl Filter {
    pub fn generate_code(
        &self,
        filter_by_attributes: &FilterByAttributes,
        entity: &Entity,
        entities: &Vec<Entity>,
    ) -> anyhow::Result<String> {
        self.content
            .iter()
            .map(|filter_body| -> anyhow::Result<String> {
                filter_body.generate_code(filter_by_attributes, entity, entities)
            })
            .collect::<anyhow::Result<Vec<String>>>()
            .map(|filter_bodies| filter_bodies.join(""))
    }
}
#[derive(Debug, Clone)]
pub enum FilterBody {
    ContentWithMSFA(ContentWithMSFA),
    FilterAttributeRep(FilterAttributeRep),
}

impl FilterBody {
    pub fn generate_code(
        &self,
        filter_by_attributes: &FilterByAttributes,
        entity: &Entity,
        entities: &Vec<Entity>,
    ) -> anyhow::Result<String> {
        match self {
            Self::ContentWithMSFA(content_with_msfa) => {
                content_with_msfa.generate_code(filter_by_attributes, entity, entities)
            }
            Self::FilterAttributeRep(filter_attribute_rep) => {
                filter_attribute_rep.generate_code(filter_by_attributes, entity, entities)
            }
        }
    }
}

#[derive(Debug, Clone)]
pub enum FilterAttributeRep {
    NamesAndTypesPresent {
        prefix: Content,
        template_attribute_name: StringWithNamingConvention,
        attribute_name_and_type_separator: Content,
        template_attribute_type: StringWithNamingConvention,
        attributes_separator: Content,
        suffix: Content,
    },
    NamesPresent {
        prefix: Content,
        template_attribute_name: StringWithNamingConvention,
        attributes_separator: Content,
        suffix: Content,
    },
    NamesTwicePresent {
        prefix: Content,
        template_attribute_name: StringWithNamingConvention,
        attribute_name_twice_separator: Content,
        attributes_separator: Content,
        suffix: Content,
    },
}

impl FilterAttributeRep {
    pub fn generate_code(
        &self,
        filter_by_attributes: &FilterByAttributes,
        entity: &Entity,
        entities: &Vec<Entity>,
    ) -> anyhow::Result<String> {
        match self {
            Self::NamesAndTypesPresent {
                prefix,
                template_attribute_name,
                attribute_name_and_type_separator,
                template_attribute_type,
                attributes_separator,
                suffix,
            } => {
                let prefix = prefix.generate_code(&entity, &entities)?;
                let attributes = filter_by_attributes
                    .iter()
                    .map(
                        |(entity_attribute_name, entity_attribute_type)| -> anyhow::Result<String> {
                            let attribute_name_and_type_separator =
                                attribute_name_and_type_separator
                                    .generate_code(&entity, &entities)?;
                            let attribute_type = template_attribute_type.to_naming_convention(
                                &entity_attribute_type.get_attribute_type(&entities)?,
                            );
                            Ok(format!(
                                "{}{}{}",
                                template_attribute_name.to_naming_convention(entity_attribute_name),
                                attribute_name_and_type_separator,
                                attribute_type
                            ))
                        },
                    )
                    .collect::<anyhow::Result<Vec<String>>>()?
                    .join(
                        attributes_separator
                            .generate_code(&entity, &entities)?
                            .as_str(),
                    );
                let suffix = suffix.generate_code(&entity, &entities)?;
                Ok(format!("{}{}{}", prefix, attributes, suffix))
            }
            Self::NamesPresent {
                prefix,
                template_attribute_name,
                attributes_separator,
                suffix,
            } => {
                let prefix = prefix.generate_code(&entity, &entities)?;
                let attributes = filter_by_attributes
                    .iter()
                    .map(|(entity_attribute_name, _)| -> anyhow::Result<String> {
                        Ok(template_attribute_name.to_naming_convention(entity_attribute_name))
                    })
                    .collect::<anyhow::Result<Vec<String>>>()?
                    .join(
                        attributes_separator
                            .generate_code(&entity, &entities)?
                            .as_str(),
                    );
                let suffix = suffix.generate_code(&entity, &entities)?;
                Ok(format!("{}{}{}", prefix, attributes, suffix))
            }
            Self::NamesTwicePresent {
                prefix,
                template_attribute_name,
                attribute_name_twice_separator,
                attributes_separator,
                suffix,
            } => {
                let prefix = prefix.generate_code(&entity, &entities)?;
                let attributes = filter_by_attributes
                    .iter()
                    .map(|(entity_attribute_name, _)| -> anyhow::Result<String> {
                        Ok(format!(
                            "{}{}{}",
                            template_attribute_name.to_naming_convention(entity_attribute_name),
                            attribute_name_twice_separator.generate_code(&entity, &entities)?,
                            template_attribute_name.to_naming_convention(entity_attribute_name)
                        ))
                    })
                    .collect::<anyhow::Result<Vec<String>>>()?
                    .join(
                        attributes_separator
                            .generate_code(&entity, &entities)?
                            .as_str(),
                    );
                let suffix = suffix.generate_code(&entity, &entities)?;
                Ok(format!("{}{}{}", prefix, attributes, suffix))
            }
        }
    }
}

#[derive(Debug, Clone)]
pub enum StringWithNamingConvention {
    KebabCase,
    LowerCamelCase,
    UpperCamelCase,
    ShoutyKebabCase,
    ShoutySnakeCase,
    SnakeCase,
    TitleCase,
    TrainCase,
}

impl StringWithNamingConvention {
    pub fn to_naming_convention(&self, value: &str) -> String {
        match self {
            Self::KebabCase => value.to_kebab_case(),
            Self::LowerCamelCase => value.to_lower_camel_case(),
            Self::UpperCamelCase => value.to_upper_camel_case(),
            Self::ShoutyKebabCase => value.to_shouty_kebab_case(),
            Self::ShoutySnakeCase => value.to_shouty_snake_case(),
            Self::SnakeCase => value.to_snake_case(),
            Self::TitleCase => value.to_title_case(),
            Self::TrainCase => value.to_train_case(),
        }
    }
}


// region: tests
#[cfg(test)]
mod tests {
    use crate::{codegen::CodeGen, models::entity::{AttributeType, Entity, ForeignKey, Semantics}, parsers::template_parser::TemplateParser};

    fn prepare_entities() -> Vec<Entity> {
         let car = Entity {
            name: "Car".to_string(),
            semantics: Semantics {
                plural: "Cars".to_string(),
            },
            attributes: vec![
                ("id".to_string(), AttributeType::Type("Uuid".to_string())),
                (
                    "name".to_string(),
                    AttributeType::Type("String".to_string()),
                ),
                (
                    "model".to_string(),
                    AttributeType::Type("String".to_string()),
                ),
                (
                    "manufacturer".to_string(),
                    AttributeType::Type("String".to_string()),
                ),
                ("year".to_string(), AttributeType::Type("i32".to_string())),
                (
                    "owner".to_string(),
                    AttributeType::ForeignKey(ForeignKey {
                        entity_name: "Person".to_string(),
                        attribute_name: "id".to_string(),
                    }),
                ),
            ]
            .into_iter()
            .collect(),
            primary_key: "id".to_string(),
            filter_by: vec![
                vec!["id".to_string()],
                vec!["name".to_string(), "model".to_string()],
                vec!["manufacturer".to_string()],
                vec!["year".to_string()],
                vec!["owner".to_string()],
            ],
            unique_attributes: vec![vec!["id".to_string()], vec!["name".to_string()]]
                .into_iter()
                .collect(),
        };
        let person = Entity {
            name: "Person".to_string(),
            semantics: Semantics {
                plural: "People".to_string(),
            },
            attributes: vec![
                ("id".to_string(), AttributeType::Type("Uuid".to_string())),
                (
                    "name".to_string(),
                    AttributeType::Type("String".to_string()),
                ),
                ("age".to_string(), AttributeType::Type("i32".to_string())),
            ]
            .into_iter()
            .collect(),
            primary_key: "id".to_string(),
            filter_by: vec![
                vec!["id".to_string()],
                vec!["name".to_string()],
                vec!["age".to_string()],
            ],
            unique_attributes: vec![vec!["id".to_string()], vec!["name".to_string()]]
                .into_iter()
                .collect(),
        };
        let entities = vec![car.clone(), person];
        entities
    }

    #[test]
    fn test_entity_code_gen() {
        let entity_model_test = std::fs::read_to_string("/Users/abdullahsabaaallil/Desktop/ene-lightspeed/template-rs/src/models/entity_name.rs").unwrap();
        let parsed_template = TemplateParser::parse_template(&entity_model_test).unwrap();
        let entities = prepare_entities();
        let generated_code = parsed_template.generate_code(&entities[0], &entities).unwrap();
        std::fs::write("test_output/entity_test.rs", generated_code.clone())
            .map_err(|e| {
                println!("Error: {:?}", e);
                e
            })
            .unwrap();
        insta::assert_debug_snapshot!(generated_code);
    }

    #[test]
    fn test_controller_code_gen() {
        let controller_test = std::fs::read_to_string("/Users/abdullahsabaaallil/Desktop/ene-lightspeed/template-rs/src/controllers/entity_controller.rs").unwrap();
        let parsed_template = TemplateParser::parse_template(&controller_test).unwrap();
        let entities = prepare_entities();
        let generated_code = parsed_template.generate_code(&entities[0], &entities).unwrap();
        std::fs::write("test_output/entity_controller_test.rs", generated_code.clone())
            .map_err(|e| {
                println!("Error: {:?}", e);
                e
            })
            .unwrap();
        insta::assert_debug_snapshot!(generated_code);
    }


}