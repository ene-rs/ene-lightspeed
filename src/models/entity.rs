use std::{collections::BTreeMap, fmt::Display};

use anyhow::bail;
use serde::{Deserialize, Serialize};

pub type AttributeName = String;
pub type FilterByAttributeNames = Vec<AttributeName>;

pub type FilterByAttributes = Vec<(AttributeName, AttributeType)>;

pub trait MostSpecificFilterAttributes {
    fn get_most_specific_filter_attribute(&self) -> (AttributeName, AttributeType);
}

impl MostSpecificFilterAttributes for &FilterByAttributes {
    fn get_most_specific_filter_attribute(&self) -> (AttributeName, AttributeType) {
        self.last().unwrap().clone()
    }
}
pub type UniqueAttributes = Vec<AttributeName>;

#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(untagged)]
pub enum AttributeType {
    Type(String),
    ForeignKey(ForeignKey),
}

impl AttributeType {
    pub fn get_attribute_type(&self, entities: &Vec<Entity>) -> anyhow::Result<String> {
        match self {
            AttributeType::Type(type_name) => Ok(type_name.clone()),
            AttributeType::ForeignKey(ForeignKey {
                entity_name,
                attribute_name,
            }) => {
                let foreign_key_entity = entities
                    .iter()
                    .find(|entity| entity.name == *entity_name)
                    .expect(&format!(
                        "Foreign key entity {} is not present in the entities",
                        entity_name
                    ));
                let foreign_key_attribute: &AttributeType = foreign_key_entity
                    .attributes
                    .get(attribute_name)
                    .expect(&format!(
                        "Foreign key attribute {} is not present in the attributes of {}",
                        attribute_name, entity_name
                    ));
                foreign_key_attribute.get_attribute_type(entities)
            }
        }
    }
}
#[derive(Serialize, Deserialize, Clone, Default)]
pub struct Entity {
    pub name: String,
    pub attributes: BTreeMap<AttributeName, AttributeType>,
    pub primary_key: AttributeName,
    pub filter_by: Vec<FilterByAttributeNames>,
    pub unique_attributes: Vec<UniqueAttributes>,
    pub semantics: Semantics,
}

impl Entity {
    /**
     * Possible constraints:
     * - All attributes used as primary key, filter by, unique attributes, or foreign keys must be present
     * - Primary key must be in the attributes
     * - If there are unique attributes, they need to be present in filter_by
     * - If an attribute is unique, it cannot be used as a sub attribute in filter_by
     * For Example:
     * name is a unique attribute
     * It is not possible to have a filter_by with: [id, name], since name is already unique
     */
    pub fn validate_request(&self, entities: Vec<Entity>) -> anyhow::Result<()> {
        let attribute_names = self.attributes.keys().collect::<Vec<&String>>();
        if !attribute_names.contains(&&self.primary_key) {
            bail!(
                "Primary key {} is not present in the attributes of {}",
                self.primary_key,
                self.name
            );
        }

        for foreign_key in self.get_foreign_keys().iter() {
            // search the foreign key in the other entities
            let foreign_key_entity = entities
                .iter()
                .find(|entity| entity.name == foreign_key.entity_name)
                .expect(&format!(
                    "Foreign key entity {} is not present in the entities",
                    foreign_key.entity_name
                ));
            if !foreign_key_entity
                .attributes
                .iter()
                .map(|(attribute_name, _)| attribute_name)
                .collect::<Vec<&String>>()
                .contains(&&foreign_key.attribute_name)
            {
                bail!(
                    "Foreign key attribute {} is not present in the attributes of {}",
                    foreign_key.attribute_name,
                    foreign_key.entity_name
                );
            }
        }

        for unique_attribute in self
            .unique_attributes
            .iter()
            .flatten()
            .collect::<Vec<&String>>()
        {
            if self.attributes.get(unique_attribute).is_none() {
                bail!(
                    "Unique attribute {} is not present in the attributes of {}",
                    unique_attribute,
                    self.name
                );
            }
        }
        for filter_by_attribute in self.filter_by.iter().flatten().collect::<Vec<&String>>() {
            if self.attributes.get(filter_by_attribute).is_none() {
                bail!(
                    "Filter by attribute {} is not present in the attributes of {}",
                    filter_by_attribute,
                    self.name
                );
            }
        }

        // Verify that the unique attributes are present in the filter_by
        for unique_attribute in self.unique_attributes.iter() {
            if !self.filter_by.contains(unique_attribute) {
                bail!(
                    "Unique attribute {:?} is not present in the filter_by of {}",
                    unique_attribute,
                    self.name
                );
            }

            let common = self
                .filter_by
                .iter()
                .filter(|filter_by| filter_by != &unique_attribute)
                .any(|filter_by| {
                    println!("{:?}", filter_by);
                    is_sub(&filter_by, &unique_attribute)
                });

            if common {
                bail!("Unique attribute {:?} is a sub attribute of another filter_by in {}. It does not make sense to have fine grained filters on unique attributes, since they're unique", unique_attribute, self.name);
            }
        }

        Ok(())
    }

    pub fn get_foreign_keys(&self) -> Vec<ForeignKey> {
        let mut foreign_keys = Vec::new();
        for (_, attribute_type) in &self.attributes {
            match attribute_type {
                AttributeType::ForeignKey(foreign_key) => {
                    foreign_keys.push(foreign_key.clone());
                }
                _ => {}
            }
        }
        foreign_keys
    }

    pub fn get_filter_by_attributes(&self) -> Vec<FilterByAttributes> {
        let mut filter_by_attributes = Vec::new();
        for filter_by in &self.filter_by {
            let mut filter_by_attribute = Vec::new();
            for attribute_name in filter_by {
                filter_by_attribute.push((
                    attribute_name.clone(),
                    self.attributes.get(attribute_name).unwrap().clone(),
                ));
            }
            filter_by_attributes.push(filter_by_attribute);
        }
        filter_by_attributes
    }
}

fn is_sub<T: PartialEq>(mut haystack: &[T], needle: &[T]) -> bool {
    if needle.len() == 0 {
        return true;
    }
    while !haystack.is_empty() {
        if haystack.starts_with(needle) {
            return true;
        }
        haystack = &haystack[1..];
    }
    false
}

#[derive(Serialize, Deserialize, Clone, Default)]
pub struct Semantics {
    pub plural: String,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct ForeignKey {
    pub entity_name: String,
    pub attribute_name: String,
}

impl Display for AttributeType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", self)
    }
}
