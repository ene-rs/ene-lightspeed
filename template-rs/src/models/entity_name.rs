// for each entity
use serde::Serialize;
use serde::Deserialize;
use uuid::Uuid;
use crate::error::Error;
use crate::models::PaginatedResult;
use serde::Serializer;
use crate::controllers::entity_name_controller::{AddEntityNamePayload, UpdateEntityNamePayload};


#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EntityName {
    
    // for each attribute
    pub attribute_name: AttributeType,
    pub attribute2_name: Attribute2Type,
    // end for each attribute
}

impl EntityName {
    pub fn new(payload: AddEntityNamePayload) -> Result<Self, Error> {
        Ok(Self {
            primary_key: payload.primary_key,
            // for each attribute
            attribute_name: payload.attribute_name,
            attribute2_name: payload.attribute2_name,
            // end for each attribute
        })
    }

    pub fn update(payload: UpdateEntityNamePayload) -> Result<Self, Error> {
        Ok(Self {
            primary_key: payload.primary_key,
            // for each attribute
            attribute_name: payload.attribute_name,
            attribute2_name: payload.attribute2_name,
            // end for each attribute
        })
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EntityNameFilterAttributes {

    // for each filter
    //    filter
    //       for each filter attribute
    filter_attribute_name: Option<FilterAttributeType>,
    filter2_attribute_name: Option<Filter2AttributeType>,
    //       end for each filter attribute
    //     end filter
    //     filter
    //        for each filter attribute
    filter_attribute_name: Option<FilterAttributeType>,
    filter2_attribute_name: Option<Filter2AttributeType>,
    //         end for each filter attribute
    //      end filter
    // end for each filter
}

impl EntityNameFilterAttributes {
    // for each filter
    //    filter
    fn is_most_specific_attribute_name_filter(&self) -> bool {
        //    for each filter attribute
        self.filter_attribute_name.is_some() && self.filter2_attribute_name.is_some()
        //     end for each filter attribute
    }
    //     end filter
    //     filter
    fn is_most_specific_attribute_name_filter(&self) -> bool {
        //    for each filter attribute
        self.filter_attribute_name.is_some() && self.filter2_attribute_name.is_some()
        //    end for each filter attribute
    }
    //     end filter
    // end for each filter

}
// end for each entity