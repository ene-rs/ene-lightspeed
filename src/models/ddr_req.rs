use serde::{Deserialize, Serialize};

use super::entity::Entity;

pub type EntityName = String;
#[derive(Serialize, Deserialize)]
pub struct DomainDrivenRequest {
    pub service_name: String,
    pub entities: Vec<Entity>,
}

impl DomainDrivenRequest {
    /**
     * Possible constraints:
     * - All attributes used as primary key, filter by, unique attributes, or foreign keys must be present
     * - Primary key must be in the attributes
     * - If there are unique attributes, they need to be present in filter_by
     * - If an attribute is unique, it cannot be used as a sub attribute in filter_by
     * For Example:
     * name is a unique attribute
     * It is not possible to have a filter_by with: [id, name], since name is already unique
     */
    pub fn validate_request(&self) -> anyhow::Result<()> {
        for entity in &self.entities {
            entity.validate_request(self.entities.clone())?;
        }
        Ok(())
    }
}
