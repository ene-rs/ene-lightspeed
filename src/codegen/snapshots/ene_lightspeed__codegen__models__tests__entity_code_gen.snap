---
source: src/codegen/models.rs
expression: generated_code
---
"\nuse serde::Serialize;\nuse serde::Deserialize;\nuse uuid::Uuid;\nuse crate::error::Error;\nuse crate::models::PaginatedResult;\nuse serde::Serializer;\nuse crate::controllers::car_controller::{AddCarPayload, UpdateCarPayload};\n\n\n#[derive(Debug, Clone, Serialize, Deserialize)]\n#[serde(rename_all = \"camelCase\")]\npub struct Car {\n    pub id: Uuid,\n    pub manufacturer: String,\n    pub model: String,\n    pub name: String,\n    pub owner: Uuid,\n    pub year: I32,\n}\n\nimpl Car {\n    pub fn new(payload: AddCarPayload) -> Result<Self, Error> {\n        Ok(Self {\n            id: payload.id,\n            id: payload.id,\n            manufacturer: payload.manufacturer,\n            model: payload.model,\n            name: payload.name,\n            owner: payload.owner,\n            year: payload.year,\n        })\n    }\n\n    pub fn update(payload: UpdateCarPayload) -> Result<Self, Error> {\n        Ok(Self {\n            id: payload.id,\n            id: payload.id,\n            manufacturer: payload.manufacturer,\n            model: payload.model,\n            name: payload.name,\n            owner: payload.owner,\n            year: payload.year,\n        })\n    }\n}\n\n#[derive(Debug, Clone, Serialize, Deserialize)]\n#[serde(rename_all = \"camelCase\")]\npub struct CarFilterAttributes {\n    id: Option<Uuid>,\n    name: Option<String>,\n    model: Option<String>,\n    manufacturer: Option<String>,\n    year: Option<I32>,\n    owner: Option<Uuid>,\n}\n\nimpl CarFilterAttributes {\n    fn is_id_filter(&self) -> bool {\n        self.id.is_some()\n    }\n    fn is_model_filter(&self) -> bool {\n        self.name.is_some() && self.model.is_some()\n    }\n    fn is_manufacturer_filter(&self) -> bool {\n        self.manufacturer.is_some()\n    }\n    fn is_year_filter(&self) -> bool {\n        self.year.is_some()\n    }\n    fn is_owner_filter(&self) -> bool {\n        self.owner.is_some()\n    }\n\n}"
